framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -381,42 +381,59 @@ public void processClassTree(ClassTree classTree) {
                 validateTypeOf(im);
             }
         }
+        checkExtendsImplements(classTree);
 
+        super.visitClass(classTree, null);
+    }
+
+    /**
+     * If "@B class Y extends @A X {}", then enforce that @B must be a subtype of @A.
+     *
+     * @param classTree class tree to check
+     */
+    protected void checkExtendsImplements(ClassTree classTree) {
+        if (classTree.getSimpleName().contentEquals("")) {
+            // Don't check extends clause on anonymous classes.
+            return;
+        }
+        AnnotatedTypeMirror classType = atypeFactory.getAnnotatedType(classTree);
+        QualifierHierarchy qualifierHierarchy = atypeFactory.getQualifierHierarchy();
         // If "@B class Y extends @A X {}", then enforce that @B must be a subtype of @A.
         // classTree.getExtendsClause() is null when there is no explicitly-written extends clause,
         // as in "class X {}". This is equivalent to writing "class X extends @Top Object {}", so
         // there is no need to do any subtype checking.
         if (classTree.getExtendsClause() != null) {
-            AnnotatedTypeMirror extendsClauseType =
-                    atypeFactory.getAnnotatedType(classTree.getExtendsClause());
-            checkExtendsOrImplementsClause(
-                    classTree, extendsClauseType, "declaration.inconsistent.with.extends.clause");
+            Set<AnnotationMirror> extendsAnnos =
+                    atypeFactory.getAnnotatedType(classTree.getExtendsClause()).getAnnotations();
+            for (AnnotationMirror classAnno : classType.getAnnotations()) {
```

Comment:
How about extracting the `for` loop into a helper method, as the same code is used below? Similar to the previous `checkExtendsOrImplementsClause` helper method.

Reply:
Maybe this method should be called `checkExtendsAndImplementsClauses` and the helper `checkExtendsOrImplementsClause`. But there are probably better names...

Reply:
I tried that, but I found it more confusing.  The helper method isn't really about extends/implements, its more like `ifNotSubtypeIssueError`. I also requires 4 arguments.